extra/modules/greenbids-real-time-data/src/main/java/org/prebid/server/hooks/modules/greenbids/real/time/data/model/data/GreenbidsInferenceData.java

```diff
@@ -0,0 +1,203 @@
+                throttlingImpMessages.add(
+                        ThrottlingMessage.builder()
+                                .browser(greenbidsUserAgent.getBrowser())
+                                .bidder(bidderName)
+                                .adUnitCode(impId)
+                                .country(countryFromIp)
+                                .hostname(hostname)
+                                .device(greenbidsUserAgent.getDevice())
+                                .hourBucket(hourBucket.toString())
+                                .minuteQuadrant(minuteQuadrant.toString())
+                                .build());
+            }
+        }
+        return throttlingImpMessages;
+    }
+
+    private static ExtImpPrebid extImpPrebid(JsonNode extImpPrebid, ObjectMapper mapper) {
+        try {
+            return mapper.treeToValue(extImpPrebid, ExtImpPrebid.class);
+        } catch (JsonProcessingException e) {
+            throw new PreBidException("Error decoding imp.ext.prebid: " + e.getMessage(), e);
+        }
+    }
+
+    private static String getCountry(String ip, File database) throws IOException, GeoIp2Exception {
+        return Optional.ofNullable(ip)
+                .map(ipAddress -> {
+                    try {
+                        final DatabaseReader dbReader = new DatabaseReader.Builder(database).build();
+                        final InetAddress inetAddress = InetAddress.getByName(ipAddress);
+                        final CountryResponse response = dbReader.country(inetAddress);
+                        final Country country = response.getCountry();
+                        return country.getName();
+                    } catch (IOException | GeoIp2Exception e) {
+                        throw new PreBidException("Failed to fetch country from geoLite DB", e);
+                    }
+                }).orElse(null);
+    }
+
+    private static String[][] convertToArray(List<ThrottlingMessage> messages) {
+        return messages.stream()
+                .map(message -> new String[]{
+                        Optional.ofNullable(message.getBrowser()).orElse(""),
+                        Optional.ofNullable(message.getBidder()).orElse(""),
+                        Optional.ofNullable(message.getAdUnitCode()).orElse(""),
+                        Optional.ofNullable(message.getCountry()).orElse(""),
+                        Optional.ofNullable(message.getHostname()).orElse(""),
+                        Optional.ofNullable(message.getDevice()).orElse(""),
+                        Optional.ofNullable(message.getHourBucket()).orElse(""),
+                        Optional.ofNullable(message.getMinuteQuadrant()).orElse("")})
+                .toArray(String[][]::new);
+    }
+
+    private static Boolean isAnyFeatureNull(String[][] throttlingInferenceRow) {
+        return Arrays.stream(throttlingInferenceRow)
+                .flatMap(Arrays::stream)
+                .anyMatch(Objects::isNull);
```

Comment:
move this convertion closer to the actual call

also move this null checks into the message building

also, are you sure that the null value is not applicable and it's required to use empty strings? looks redundant

Reply:
have moved to FilterService, where `onnxModelRunner.runModel(throttlingInferenceRows);`  is called